src/infra/src/table/migration/m20241213_102300_create_destinations_table.rs

```diff
@@ -0,0 +1,179 @@
+// Copyright 2024 OpenObserve Inc.
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+use sea_orm_migration::prelude::*;
+
+#[derive(DeriveMigrationName)]
+pub struct Migration;
+
+const DESTINATIONS_TEMPLATES_FK: &str = "destinations_templates_fk";
+const DESTINATIONS_TEMPLATE_ID_DESTINATION_ID_IDX: &str =
+    "destinations_template_id_destination_id_idx";
+
+#[async_trait::async_trait]
+impl MigrationTrait for Migration {
+    async fn up(&self, manager: &SchemaManager) -> Result<(), DbErr> {
+        manager
+            .create_table(create_destinations_table_statement())
+            .await?;
+        manager
+            .create_index(create_destinations_template_id_destination_id_idx_stmnt())
+            .await?;
+        Ok(())
+    }
+
+    async fn down(&self, manager: &SchemaManager) -> Result<(), DbErr> {
+        manager
+            .drop_index(
+                Index::drop()
+                    .name(DESTINATIONS_TEMPLATE_ID_DESTINATION_ID_IDX)
+                    .to_owned(),
+            )
+            .await?;
+        manager
+            .drop_table(Table::drop().table(Destinations::Table).to_owned())
+            .await?;
+        Ok(())
+    }
+}
+
+/// Statement to create the destinations table.
+fn create_destinations_table_statement() -> TableCreateStatement {
+    Table::create()
+        .table(Destinations::Table)
+        .if_not_exists()
+        // The ID is 27-character human readable KSUID.
+        .col(
+            ColumnDef::new(Destinations::Id)
+                .char_len(27)
+                .not_null()
+                .primary_key(),
+        )
+        .col(ColumnDef::new(Destinations::Org).string_len(100).not_null())
+        .col(ColumnDef::new(Destinations::Name).string_len(256).not_null())
+        .col(ColumnDef::new(Destinations::Module).string_len(10).not_null())
+        .col(ColumnDef::new(Destinations::TemplateId).char_len(27).null())
+        .col(ColumnDef::new(Destinations::PipelineId).string_len(27).null()) // pipeline id created w/ varchar
+        .col(ColumnDef::new(Destinations::Type).json().not_null())
```

Comment:
Nit:

Would it possibly make more sense to store type as an `i16` or some sort of string?

Honestly I don't think it makes a huge difference one way or another. But if we store this as `json` then every time we store a value here we'll have to store `"http"` (including the double quotes) instead of just `http` since `"http"` is valid JSON but `http` (without the double quotes is not).

Reply:
I realize you've got a lot of PRs based off of this one, so if changing this is going to be a hassle I wouldn't worry about it. Besides, if for some reason one day we did actually care about storing `type` as an `i16` or `char(10)` or some other column type like that in the future, then it wouldn't be hard to migrate that. 

Regardless of whether this column is `json`, `i16`, `char(10)`, etc... the column type is always going to permit "bad" values and so we'll always have to implement some sort of "failable conversion" (ie the `TryFrom` trait) to convert from the DB column type to the actual Rust `DestinationType`. So the choice between those column types doesn't seem significant to me

Reply:
I just started looking at the follow up to this PR and it looks like "type" can actually be a struct with fields (rather than simply an enum). I didn't realize that before, so if that's the case please ignore my previous comments